### PR TITLE
RFC: Recursive cp

### PR DESCRIPTION
**Pull request by ghost**
Addresses #10434, went with keyword argument as proposed by @simonster to be in line with `rm`.  Tests included.

Late-night recreational coding, second opinions appreciated.

### CODE REVIEW

File: base/file.jl [hunk 1]
```diff
@@ -70,7 +70,16 @@ end
...
-cp(src::AbstractString, dst::AbstractString) = FS.sendfile(src, dst)
+function cp(src::AbstractString, dst::AbstractString; recursive::Bool=false)
+    if islink(src) || !isdir(src)
+        FS.sendfile(src, dst)
+    elseif recursive
+        mkdir(dst)
+        for p in readdir(src)
+            cp(joinpath(src, p), joinpath(dst, p), recursive=recursive)
+        end
+    end
```
Comment (simonster) [inline base/file.jl]: We should probably have an `else` here that throws an error if the user tries to copy a directory without `recursive`.

Comment (ninjin) [inline base/file.jl]: Agreed, should we create a new error for this?  Python throws `IsADirectoryError`.

**Comment by ivarne** [comment]
`ArgumentError` seems appropriate enough. We don't yet have detailed exception hierarchies in Julia.

We should also have a suggestion to use the recursive flag in the error message. Otherwise we'll get bug reports that cp can't copy directories, which is a waste of our users time (and our own when we have to answer ).

**Comment by ninjin** [comment]
Pushed new version that raises `ArgumentError` when copying directories without `recursive`.

File: base/file.jl [hunk 2]
```diff
@@ -70,7 +70,19 @@ end
...
+    if islink(src) || !isdir(src)
+        FS.sendfile(src, dst)
+    elseif recursive
+        mkdir(dst)
+        for p in readdir(src)
+            cp(joinpath(src, p), joinpath(dst, p), recursive=recursive)
+        end
+    else
+        throw(ArgumentError(string("'$src' is a directory. ",
+            "To copy, set the `recursive` keyword argument to true.")))
```
Comment (ivarne) [inline base/file.jl]: How about
```
'$src' is a directory.
Use `cp(src, dst, recursive=true)` to copy directories recursively.
```
I don't like to rely on users understanding terminology.

**Comment by ninjin** [comment]
Travis Linux i686 failure appears unrelated, OS X timed out.  All others passed.

File: base/file.jl (hunk 2 above)
Comment (ninjin) [inline base/file.jl]: Agreed, updating and pushing.

**Comment by ninjin** [comment]
OS X timed out again.

**Comment by tkelman** [comment]
Good to see another PR from you @ninjin!

**Comment by ninjin** [comment]
@tkelman: I am recovering, hopefully I can take a stab at extending the Travis integration shortly.

**Comment by tkelman** [comment]
this is probably worth mentioning in the docs, huh

**Comment by ivarne** [comment]
Docs in https://github.com/JuliaLang/julia/commit/45e5fa15c03571c3ffb1bead9c03fa3a7ac30f09

**Comment by tkelman** [comment]
thanks @ivarne

**Pull request merged**